slides/k8s/build-with-docker.md

```diff
@@ -0,0 +1,156 @@
+
+- To access the Docker Engine, we will mount the Docker socket in our container
+
+- After building the image, we will push it to our self-hosted registry
+
+---
+
+## Resource specification for our builder pod
+
+.small[
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: build-image
+spec:
+  restartPolicy: OnFailure
+  containers:
+  - name: docker-build
+    image: docker
+    env:
+    - name: REGISTRY_PORT
+      value: "`3XXXX`"
+    command: ["sh", "-c"]
+    args:
+    - |
+      apk add --no-cache git &&
+      mkdir /workspace &&
+      git clone https://github.com/jpetazzo/container.training /workspace &&
+      docker build -t localhost:$REGISTRY_PORT/worker /workspace/dockercoins/worker &&
+      docker push localhost:$REGISTRY_PORT/worker
+    volumeMounts:
+    - name: docker-socket
+      mountPath: /var/run/docker.sock
+  volumes:
+  - name: docker-socket
+    hostPath:
+      path: /var/run/docker.sock
+```
+]
+
+---
+
+## Breaking down the pod specification (1/2)
+
+- `restartPolicy: OnFailure` prevents the build from running in an infinite lopo
+
+- We use the `docker` image (so that the `docker` CLI is available)
+
+- We rely on the fact that the `docker` image is based on `alpine`
+
+  (which is why we use `apk` to install `git`)
+
+- The port for the registry is passed through an environment variable
+
+  (this avoids to repeat it in the specification, which would be error-prone)
+
+.warning[The environment variable has to be a string, so the `"` are mandatory!]
+
+---
+
+## Breaking down the pod specification (2/2)
+
+- The volume `docker-socket` is declared with a `hostPath`, indicating a bind-mount
+
+- It is then mounted in the container onto the default Docker socket path
+
+- We show a creative way to specify the commands to run in the container:
+
+  - the command executed will be `sh -c <args>`
+
+  - `args` is a list of strings
+
+  - `|` is used to pass a multi-line string in the YAML file
+
+---
+
+## Running our pod
+
+- Let's try this out!
+
+.exercise[
+
+- Check the port used by our self-hosted registry:
+  ```bash
+  kubectl get svc registry
+  ```
+
+- Edit `~/container.training/k8s/docker-build.yaml` to put the port number
+
+- Schedule the pod by applying the resource file:
+  ```bash
+  kubectl apply -f ~/container.training/k8s/docker-build.yaml
+  ```
+
+- Watch the logs:
+  ```bash
+  stern build-image
+  ```
+
+]
+
+---
+
+## What's missing?
+
+What do we need to change to make this production-ready?
+
+- Build from a long-running container (e.g. a `Deployment`) triggered by web hooks
+
+  (the payload of the web hook could indicate the repository to build)
+
+- Build a specific branch or tag; tag image accordingly
+
+- Handle repositories where the Dockerfile is not at the root
+
+  (or containing multiple Dockerfiles)
+
+- Expose build logs so that troubleshooting is straightforward
+
+--
+
+🤔 That seems like a lot of work!
+
+--
+
+That's why systems like Docker Hub are helpful.
```

Comment:
"systems" is confusing here, to people who may think of "Docker Hub" as "that place I push images" and nothing else. Maybe explain a little more how it's a "system"?

Reply:
Absolutely.

I'm replacing this line with:

That's why services like Docker Hub (with [automated builds](https://docs.docker.com/docker-hub/builds/)) are helpful.

I hope that's better.